src/common/middleware/call-api.js

```diff
@@ -17,39 +17,67 @@ export default defaults => () => next => async action => {
   validateDefaults(defaults);
   validateSettings(settings);
 
-  const [ requestType, successType, failureType ] = settings.types;
+  const requestType = settings.types ? settings.types[0] : null;
+  const successType = settings.types ? settings.types[1] : null;
+  const failureType = settings.types ? settings.types[2] : null;
   const resource = defaults.endpoint + settings.path;
   const optionsWithCsrfToken = withCsrfToken(defaults.csrfToken, settings.options);
   const createAction = createActionWith.bind({}, action);
 
   // Immediately dispatch request action
-  next(createAction({ type: requestType }));
+  if (requestType) {
+    next(createAction({ type: requestType }));
+  }
 
-  try {
-    const response = await fetch(resource, optionsWithCsrfToken);
-    await checkStatus(response);
-    const result = await response.json();
+  return fetch(resource, optionsWithCsrfToken)
+    .then((response) => {
+      return checkStatus(response).then(() => {
+        return response.json().then((result) => {
+          if (result.status !== 'success') {
+            throw getError(response, result);
+          }
 
-    if (result.status !== 'success') {
-      throw getError(response, result);
-    }
+          if (successType) {
+            return next(createAction({
+              type: successType,
+              payload: {
+                response: result
+              }
+            }));
+          }
 
-    return next(createAction({
-      type: successType,
-      payload: {
-        response: result
-      }
-    }));
-  }
-  catch(error) {
-    return next(createAction({
-      type: failureType,
-      payload: {
-        error: error
-      },
-      error: true
-    }));
-  }
+          return Promise.resolve(result);
+        });
+      }, (error) => {
+        if (failureType) {
+          return next(createAction({
+            type: failureType,
+            payload: {
+              error: error
+            },
+            error: true
+          }));
+        }
+
+        error.action = action;
+
+        return Promise.reject(error);
+      }, (error) => {
+        if (failureType) {
+          return next(createAction({
+            type: failureType,
+            payload: {
+              error: error
+            },
+            error: true
+          }));
+        }
+
+        error.action = action;
+
+        return Promise.reject(error);
```

Comment:
In various parts of our codebase we have a few API calls that do not result in actions being dispatched, but instead the resolution of the call is used to influence other actions.

A good example is `createWebhook`, which is now a fully-fledged action creator in this pull request. When this API call returns an error result, the outcome is handled by the `addRepos` action creator and stops the application redirecting the user back to the dashboard. There is currently no need for `createWebhook` to dispatch more actions.

My proposal is to allow the call-api middleware to return resolved or rejected promises so that `dispatch({ [CALL_API]: ... })` to be used with `Promise.all`, allowing API calls to be batched. There is [some precedence](https://github.com/acdlite/redux-promise/blob/master/src/index.js#L20) for middleware doing this, although I admit it feels like "making it up while I'm going along".

Let's see if we can come up with a better solution together, if there is one.